File: frontend/src/lib/alerts.ts
import type { Alert, SeverityLevel } from "@/types";

export const SEVERITY_RANK: Record<SeverityLevel, number> = {
  low: 0,
  moderate: 1,
  high: 2,
  very_high: 3,
  critical: 4,
};

/** Most-severe active alert, tie-broken by most recent. */
export function topAlert(alerts: Alert[]): Alert | null {
  if (alerts.length === 0) return null;
  return [...alerts].sort(
    (a, b) =>
      SEVERITY_RANK[b.severity] - SEVERITY_RANK[a.severity] ||
      new Date(b.issued_at).getTime() - new Date(a.issued_at).getTime()
  )[0];
}

export function alertSourceLabel(source: Alert["source"]): string {
  switch (source) {
    case "imd":
      return "IMD";
    case "sachet":
      return "SACHET";
    case "simulator":
      return "Disaster Simulator";
    case "admin":
      return "Command Center";
    default:
      return "Alert";
  }
}

export function alertRelativeTime(iso: string): string {
  const secs = Math.round((Date.now() - new Date(iso).getTime()) / 1000);
  if (secs < 60) return "just now";
  const mins = Math.round(secs / 60);
  if (mins < 60) return `${mins}m ago`;
  const hrs = Math.round(mins / 60);
  if (hrs < 24) return `${hrs}h ago`;
  return `${Math.round(hrs / 24)}d ago`;
}

File: android/app/src/main/java/com/drishti/citizen/data/model/Alerts.kt
package com.drishti.citizen.data.model

import com.drishti.citizen.core.ui.component.BadgeTone
import com.drishti.citizen.data.remote.dto.AlertDto
import java.time.Instant
import kotlin.math.roundToLong

/** Ports `frontend/src/lib/alerts.ts`. */
object Alerts {

    private val SEVERITY_RANK = mapOf(
        "low" to 0,
        "moderate" to 1,
        "high" to 2,
        "very_high" to 3,
        "critical" to 4,
    )

    /** Most-severe alert, ties broken by most recent `issued_at`. */
    fun topAlert(alerts: List<AlertDto>): AlertDto? =
        alerts.maxWithOrNull(
            compareBy({ SEVERITY_RANK[it.severity] ?: 0 }, { it.issuedAt }),
        )

    fun sourceLabel(source: String): String = when (source) {
        "imd" -> "IMD"
        "sachet" -> "SACHET"
        "simulator" -> "Disaster Simulator"
        "admin" -> "Command Center"
        else -> "Alert"
    }

    fun severityLabel(severity: String): String = severity.replace('_', ' ')

    /** Chip colour for the authority-messages timeline (web: navy / warn / danger). */
    fun severityTone(severity: String): BadgeTone = when (severity) {
        "low" -> BadgeTone.INFO
        "moderate" -> BadgeTone.WARN
        else -> BadgeTone.DANGER
    }

    fun relativeTime(iso: String, nowEpochMs: Long = System.currentTimeMillis()): String {
        val issuedAt = runCatching { Instant.parse(iso).toEpochMilli() }.getOrNull() ?: return "—"
        val secs = (nowEpochMs - issuedAt) / 1000.0
        if (secs < 60) return "just now"
        val mins = (secs / 60).roundToLong()
        if (mins < 60) return "${mins}m ago"
        val hrs = (mins / 60.0).roundToLong()
        if (hrs < 24) return "${hrs}h ago"
        return "${(hrs / 24.0).roundToLong()}d ago"
    }
}
